generate_ci.py

```diff
@@ -0,0 +1,416 @@
+
+
+for name, r in repos.items():
+    directory = name[:name.index("/")]
+    shutil.rmtree(directory, ignore_errors=True)
+    os.makedirs(directory, exist_ok=True)
+
+shutil.rmtree(openshift_release[:openshift_release.index("/")], ignore_errors=True)
+os.makedirs(openshift_release[:openshift_release.index("/")], exist_ok=True)
+
+for name, r in repos.items():
+    clone_repository(name)
+
+clone_repository(openshift_release, def_branch="master")
+
+
+@dataclasses.dataclass
+class Image:
+    name: str
+    env_name: str
+    path: str
+
+    def __lt__(self, other):
+        return self.name < other.name
+
+
+for name, r in repos.items():
+
+    org = name[:name.index('/')]
+    repo_name = name[name.index('/') + 1:]
+
+    branches = supported_branches
+    if r.get("branch") is not None and r["branch"].get("additional") is not None:
+        branches = supported_branches + r["branch"]["additional"]
+
+    # TODO: for now we're reconciling only supported branches, eventually we need to reconcile the entire
+    #  directory to make sure that we delete periodic jobs from unsupported branches
+    for supported_branch in branches:
+        run(f"rm -rf {openshift_release_ci_config}/{name}/*{supported_branch}*.yaml", stdout=PIPE, stderr=PIPE,
+            universal_newlines=True, shell=True)
+
+    for ov in openshift_versions:
+        for supported_branch in branches:
+
+            if supported_branch == "release-next":
+                promotion_name = "knative-nightly"
+            else:
+                promotion_name = supported_branch.replace("release", "knative")
+
+            print(f"=== {name} ===")
+            checkout_branch(name, supported_branch)
+
+            test_targets = get_test_targets(name, r)
+            print("---> Matching targets", test_targets)
+
+            build_image = "openshift/ci-operator/build-image"
+            print("---> Build image", build_image)
+
+            images = get_images(path="openshift/ci-operator/knative-images", name=name, r=r)
+            print("---> Images", images)
+            test_images = get_images(path="openshift/ci-operator/knative-test-images", name=name, r=r,
+                                     prefix_ctx="test-")
+            print("---> Test images", test_images)
+
+            mirroring_lines = []
+            for img in images + test_images:
+                frm = f"{openshift_release_registry}/{promotion_name}:{img.name}"
+                to = f"{quay_registry}/{img.name}:{promotion_name}"
+                mirroring_lines.append(f"{frm} {to}\n")
+
+            file_name = f"mapping_{promotion_name}_{repo_name}_quay"
+            with open(f"{openshift_release_image_mirroring}/{file_name}", "w") as f:
+                f.writelines(mirroring_lines)
+
+            images_field = ""
+            dependencies = ""
+            for img in images + test_images:
+                images_field += f"""
+- dockerfile_path: openshift/{img.path}/Dockerfile
+  to: {img.name}"""
```

Comment:
I've notice this "auto-fix" PRs.  https://github.com/openshift/release/commit/eb4d36ee1d223e13d75dfbd2d823778a2aedc708
IMO it's driven from whatever is detected in Dockerfile. https://github.com/openshift/knative-eventing/blob/release-next/openshift/ci-operator/knative-images/apiserver_receive_adapter/Dockerfile

The most simple solution would be to either define add similar builder stream to `base_images` and then`from:` here:
```
- dockerfile_path: openshift/ci-operator/knative-images/kafka-controller/Dockerfile
  from: <builder_stream_name>
  to: knative-eventing-kafka-broker-kafka-controller
```

It can be optimized a bit to to pre-build those binaries at once and copy them over to images. And stop running `make install test-install` for every image build.